### PR TITLE
Allow debugger reconnection

### PR DESCRIPTION
**Pull request by TianlongLiang**
Waiting for a new debugger connection once disconnected

1. When receiving a detach command
2. When the client socket is closed(for example, lldb process is killed)

### CODE REVIEW

File: core/iwasm/libraries/debug-engine/handler.c [hunk 1]
```diff
@@ -777,3 +777,13 @@ handle____request(WASMGDBServer *server, char *payload)
         handle_free(server, args);
     }
 }
+
+void
+handle_detach_request(WASMGDBServer *server, char *payload)
+{
+    if(payload != NULL){
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/handler.c]:
```suggestion
    if (payload != NULL){
```

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 2]
```diff
@@ -987,6 +1016,38 @@ wasm_debug_instance_interrupt_all_threads(WASMDebugInstance *instance)
...
+            /* Resume all threads since remote debugger detached*/
+            os_mutex_lock(&exec_env->wait_lock);
+            exec_env->current_status->running_status = STATUS_RUNNING;
+            os_cond_signal(&exec_env->wait_cond);
+            os_mutex_unlock(&exec_env->wait_lock);
+        }
+        exec_env = bh_list_elem_next(exec_env);
+    }
+
+    /* relaunch, and stop the thread*/
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:1044]:
```suggestion
    /* relaunch, and stop the thread */
```

Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/debug_engine.c:1044]: does it really stop the thread? Seems the thread still exists, just block on accept again?
Had better change to "relaunch, accept new debug connection"

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 3]
```diff
@@ -54,10 +54,22 @@ allocate_instance_id()
     return id;
 }
 
+static bool
+should_run(WASMDebugControlThread *control_thread)
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:58]: how about rename to `is_running`?

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 4]
```diff
@@ -54,10 +54,22 @@ allocate_instance_id()
...
+{
+    return control_thread->status == RUNNING;
+}
+
 static bool
 should_stop(WASMDebugControlThread *control_thread)
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:64]: how about rename to `is_stopped`?

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 5]
```diff
@@ -54,10 +54,22 @@ allocate_instance_id()
...
 {
-    return control_thread->status != RUNNING;
+    return control_thread->status == STOPPED;
+}
+
+static bool
+should_detach(WASMDebugControlThread *control_thread)
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:70]: how about rename to `is_detached`?

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 6]
```diff
@@ -109,73 +121,90 @@ control_thread_routine(void *arg)
...
-                     * exec_env_list.len without lock */
-                    if (debug_inst->cluster->exec_env_list.len != 1) {
-                        debug_inst->stopped_thread = NULL;
-                        /* The exiting thread may wait for the signal */
-                        os_cond_signal(&debug_inst->wait_cond);
-                        os_mutex_unlock(&control_thread->wait_lock);
-                        continue;
+        /* wait lldb client to connect */
+        if (!wasm_gdbserver_accept(control_thread->server)) {
+            LOG_ERROR("Failed while connecting debugger\n");
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:134]:
```suggestion
            LOG_ERROR("Failed while accepting debugger connection\n");
```

File: core/iwasm/libraries/debug-engine/debug_engine.c [hunk 7]
```diff
@@ -109,73 +121,90 @@ control_thread_routine(void *arg)
...
+            wasm_runtime_free(control_thread->server);
```
Comment (xujuntwt95329) [inline core/iwasm/libraries/debug-engine/debug_engine.c:135]: `control_thread->server` will be freed in `wasm_debug_control_thread_destroy`, if you free it here, there may be double free issue

Comment (TianlongLiang) [reply core/iwasm/libraries/debug-engine/debug_engine.c:135]: Should I delete this line and another line above which free `control_thread->server` too? Or should I set `control_thread->server` to NULL after free?
above line:
```c
if (!wasm_gdbserver_listen(control_thread->server)) {
    LOG_ERROR("Failed while listening for debugger\n");
    wasm_runtime_free(control_thread->server);
    return NULL;
}
```

File: core/iwasm/libraries/debug-engine/debug_engine.c (hunk 3 above)
Comment (wenyongh) [reply core/iwasm/libraries/debug-engine/debug_engine.c:58]: Had better rename to `is_thread_running`, `is_thread_stopped`, and `is_thread_detached` to make the code more clear.

File: core/iwasm/libraries/debug-engine/debug_engine.c (hunk 7 above)
Comment (xujuntwt95329) [reply core/iwasm/libraries/debug-engine/debug_engine.c:135]: Should delete these lines since the space for server will be freed during exiting

**Review by xujuntwt95329 — COMMENTED** [review]
LGTM

**Pull request merged**